_studio/mfx_lib/decode/vp9/src/mfx_vp9_dec_decode_hw.cpp

```diff
@@ -854,10 +869,24 @@ mfxStatus VideoDECODEVP9_HW::DecodeFrameCheck(mfxBitstream *bs, mfxFrameSurface1
     if (UMC::UMC_OK != videoInfo.Init(m_vPar.mfx.FrameInfo.Width, m_vPar.mfx.FrameInfo.Height, cf, m_frameInfo.bit_depth))
         return MFX_ERR_MEMORY_ALLOC;
 
-    if (UMC::UMC_OK != m_FrameAllocator->Alloc(&currMid, &videoInfo, 0))
+#ifndef MFX_VA_LINUX
+    UMC::Status umc_sts = m_FrameAllocator->Alloc(&currMid, &videoInfo, 0);
+#else
+    UMC::Status umc_sts = m_FrameAllocator->Alloc(&currMid, &videoInfo, 1); // 1 means reallocation allowed
+    if (UMC::UMC_ERR_NOT_ENOUGH_BUFFER == umc_sts && m_adaptiveMode)
     {
-        return MFX_ERR_MEMORY_ALLOC;
+        mfxFrameSurface1 *new_surf = m_FrameAllocator->GetSurfaceByIndex(currMid);
```

Comment:
this work of surface preparation should be hidden inside  function which accepts mid as input

Reply:
How? m_FrameAllocator is decoder object property?

Reply:
You have several options:
anonymous function which accepts allocator as well
another private interface function, this one looks pretty logical